clusterloader2/pkg/measurement/common/probes/probes.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package probes
+
+import (
+	"fmt"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/perf-tests/clusterloader2/pkg/flags"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/config"
+	"k8s.io/perf-tests/clusterloader2/pkg/errors"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework/client"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/prometheus"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+)
+
+const (
+	name            = "Probes"
+	probesNamespace = "probes"
+
+	manifestGlob = "$GOPATH/src/k8s.io/perf-tests/clusterloader2/pkg/measurement/common/probes/manifests/*.yaml"
+
+	checkProbesReadyInterval = 30 * time.Second
+	checkProbesReadyTimeout  = 5 * time.Minute
+
+	currentProbesMetricsVersion = "v1"
+)
+
+var probesEnabled bool
+
+func init() {
+	flags.BoolEnvVar(&probesEnabled, "enable-probes", "ENABLE_PROBES", false, "Whether to run the probes in the clusterloader2 tests.")
+	measurement.Register(name, createProbesMeasurement)
+}
+
+func createProbesMeasurement() measurement.Measurement {
+	return &probesMeasurement{
+		probeNameToPrometheusQueryTmpl: map[string]string{
+			"in_cluster_network_latency": "quantile_over_time(0.99, probes:in_cluster_network_latency:histogram_quantile[%v])",
+		},
+	}
+}
+
+type probesMeasurement struct {
+	// probeNameToPrometheusQueryTmpl defines a config of this measurement. Updating the config in the
+	// createProbesMeasurement method is the only place in go code that needs to be changed while
+	// adding a new probe.
+	// Each query template should accept a single %v placeholder corresponding to the query window
+	// length. See the 'in_cluster_network_latency' as an example.
+	probeNameToPrometheusQueryTmpl map[string]string
+
+	framework        *framework.Framework
+	replicasPerProbe int
+	templateMapping  map[string]interface{}
+	nServiceMonitors int
+	startTime        time.Time
+	mutex            sync.Mutex
```

Comment:
What is mutex protecting?

i think that in all other measurements there was silent assumption that the Execute method is not thread-safe and we don't do any locking. So if this was the motivation here, I would suggest removing this for now (just for consistency).